sentry_sdk/integrations/opentelemetry/integration.py

```diff
@@ -0,0 +1,163 @@
+
+if TYPE_CHECKING:
+    from typing import Dict
+
+
+INSTRUMENTED_CLASSES = {
+    # A mapping of packages to (original class, instrumented class) pairs. This
+    # is used to post-instrument any classes that were imported before OTel
+    # instrumentation took place.
+    "fastapi": (
+        "fastapi.FastAPI",
+        "opentelemetry.instrumentation.fastapi._InstrumentedFastAPI",
+    ),
+    "flask": ("flask.Flask", "opentelemetry.instrumentation.flask._InstrumentedFlask"),
+}
+
+
+class OpenTelemetryIntegration(Integration):
+    identifier = "opentelemetry"
+
+    @staticmethod
+    def setup_once():
+        # type: () -> None
+        logger.warning(
+            "[OTel] Initializing highly experimental OpenTelemetry support. Use at your own risk."
+        )
+
+        original_classes = _record_unpatched_classes()
+
+        try:
+            distro = _load_distro()
+            distro.configure()
+            _load_instrumentors(distro)
+        except Exception:
+            logger.exception("[OTel] Failed to auto-initialize OpenTelemetry")
+
+        _patch_remaining_classes(original_classes)
+        _setup_sentry_tracing()
+
+        logger.debug("[OTel] Finished setting up OpenTelemetry integration")
+
+
+def _record_unpatched_classes():
+    # type: () -> Dict[str, type]
+    """
+    Keep references to classes that are about to be instrumented.
+
+    Used to search for unpatched classes after the instrumentation has run so
+    that they can be patched manually.
+    """
+    installed_packages = _get_installed_modules()
+
+    original_classes = {}
+
+    for package, (orig_path, _) in INSTRUMENTED_CLASSES.items():
+        if package in installed_packages:
+            try:
+                original_cls = _import_by_path(orig_path)
+            except (AttributeError, ImportError):
+                logger.debug("[OTel] Failed to import %s", orig_path)
+                continue
+
+            original_classes[package] = original_cls
+
+    return original_classes
+
+
+def _patch_remaining_classes(original_classes):
+    # type: (Dict[str, type]) -> None
+    """
+    Best-effort attempt to patch any uninstrumented classes in sys.modules.
+
+    This enables us to not care about the order of imports and sentry_sdk.init()
+    in user code. If e.g. the Flask class had been imported before sentry_sdk
+    was init()ed (and therefore before the OTel instrumentation ran), it would
+    not be instrumented. This function goes over remaining uninstrumented
+    occurrences of the class in sys.modules and patches them.
+
+    Since this is looking for exact matches, it will not work in some scenarios
+    (e.g. if someone is not using the specific class explicitly, but rather
+    inheriting from it). In those cases it's still necessary to sentry_sdk.init()
+    before importing anything that's supposed to be instrumented.
+    """
+    for module_name, module in sys.modules.copy().items():
```

Comment:
I don't understand the outermost loop here, why do we need to go through `sys.modules` and then `original_classes` in each of them? Can you explain with a simple flask example?

Reply:
So imagine a user has something like this:
```
# app.py

from flask import Flask
import sentry_sdk

sentry_sdk.init(...)  # potel enabled
app = Flask()
```
The OTel autoinstrumentation runs during `sentry_sdk.init()`, patching the `Flask` class with its own `_InstrumentedFlask`, but since the user imported `Flask` before, in `app.py` it is still the old unpatched `Flask`, and so is `app`.

The loop goes through all modules loaded so far via `sys.modules` and does the following for each of the original classes:
* Tries reimporting them to see if what we import is instrumented or not. This is used to check if the autoinstrumentation was actually successful in the first place, because if not, we don't want to patch any leftover unpatched classes. (This check needs to be moved out of this loop since this is something that needs to only be checked once for a class, I will do that.)
* After having verified that the patching was successful, it then goes through the `vars` of each `sys.module` and checks whether the original type is in scope, and if so, replaces it with the patched type.

So in the above example, it would find the `app.py` module in `sys.modules`, it would check it for occurrences of the original `Flask` type, and replace them with `_InstrumentedFlask`.